### ISSUE TITLE
Add valign parameter for path_text()

### ISSUE THREAD

**Issue by alidaf** [revarbat/BOSL2#1022]
Text on circular paths but different directions exhibit kerning differences associated with the baseline being fixed in one position, even when compensating for the text height. Adjusting the kerning with the kern parameter is impractical when dealing with lots of text that changes frequently (using openSCAD code as a customisable template)

The standard text function has valign parameter that is useful for shifting the reference and this may be helpful in preventing letters bunching up at the top or bottom when trying to maintain the same effective radius of the path.

**Example Code**
```
include <BOSL2/std.scad>

ang1 = [0, 360];
ang2 = [360, 0];

tsize = 10;

r = 30;
r1 = r;
r2 = r - tsize; // Compensating for font size

path1 = path3d(arc(r = r1, angle = ang1));
path2 = path3d(arc(r = r2, angle = ang2));

%circle(r);

rotate([0, 0, 90])
path_text(path1,
         "Bottom Text",
          size = tsize,
          normal = UP,
          thickness = 5,
          textmetrics = true,
          center = true);

rotate([0, 0, -90])
path_text(path2,
         "Top Text",
          size = tsize,
          normal = UP,
          thickness = 5,
          textmetrics = true,
          center = true);
```

**Comment by adrianVmariano** [comment]
I can pass valign through to text() but I'm not sure how useful that actually is (or really how useful valign is period) because it screws up the text alignment.  Nobody writes text that looks like this with valign="top":

![image](https://user-images.githubusercontent.com/26014420/212767149-77b11339-0ab0-4b87-93fd-e83d8474ec1a.png)

**Comment by alidaf** [comment]
I use valign = center a lot for aligning straight lines of text within a text box so disagree with it being useless. My example is based on corporate badges and coins where text doesn't all flow clockwise, but is presented for readability. Look at a round coin and the monetary value is nearly always at the bottom, right way up. Granted the value could be placed independently of the circumferential text but many corporate logos have contra-flowing circumferential text. Just do a search on round corporate logos and you'll see many.

Another problem that your example shows is that, with the reference being what looks like bottom left, a lot of letters don't align with the radii particularly aesthetically. Look at the "T" in the bottom text. The radial line passes through the left hand edge of the upper limb making the whole letter seem misaligned. It should pass through the centre of the letter for better alignment.

**Comment by adrianVmariano** [comment]
Really?  I challenge you to find an example of typography anywhere in which the letters are aligned at the top (as in the example above) or even the center.  It looks absurd.  Letters must be aligned at their baseline or it's just incorrect typography.   The problem is that each letter is placed separately.  I think when you use valign with text() it aligns the entire text block as you specify.   I suppose there might be cases where this would work OK, where all the letters are upper case and happen to have the same top point, but that's not generic enough to add the feature.   

Something is weird with the t's in Bottom.  I don't know why they aren't oriented the right way.  It has nothing to do with valign.  It's always wrong.  It's like "tt" is being treated as a ligature.

**Comment by alidaf** [comment]
I think you are confusing how the valign works in the text() function. Look at https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/Text to see how it works there. It doesn't treat each letter separately as your example seems to be doing.

**Comment by adrianVmariano** [comment]
I think you are confused about how path_text() works.  Each letter is positioned individually.  You can see the affect above.  If I pass valign through, then each letter is *independently* aligned, producing a bogus result.

**Comment by alidaf** [comment]
Actually I'm not, which is why I was making this request with explicit reference to the text() function. I simply requested a feature to hopefully make it more consistent with the text() function. The implementation of path_text() by considering each letter as independent is irrelevant to me. If you think it can't be done because of the how the path_text() function has been implemented then that's fine. It's your prerogative to implement your code, your way.  but I fail to understand why you seem so annoyed. In any case I'll close this to prevent any further annoyance.

**Comment by adrianVmariano** [comment]
Hmmm.  Somehow I got into my head that you were asking for a passthrough to the text() valign.  Maybe it has to do with knowing the internals.  

So here's what happens if I try to implement this using the textmetrics height data.  I feel like the spacing still looks bad.  There's also the problem that you can't control where the baseline of the text lands, maybe OK, maybe not?  I guess you can check textmetrics yourself to get it where you want it, but it seems fiddly.  

![image](https://user-images.githubusercontent.com/26014420/213013611-1eac7f59-1ebe-4d69-8484-56a11e0c8685.png)

I wonder whether implementing scaling like text() has---where the letters are spaced apart proportionally to their size---would actually work better, though it might still benefit from fiddling.

**Comment by adrianVmariano** [comment]
I think the spacing problem with top alignment is that the spacing is set from a line that's above all the non-ascending characters, so they get spaced apart extra far.   This looks basically OK:

![image](https://user-images.githubusercontent.com/26014420/213014834-0e10f7e8-5fcd-4310-aa19-f1392de2fbd6.png)

(Though there seems to be some kind of bug in the computation of the normals...why is the m tilting to the right?)

**Comment by adrianVmariano** [comment]
![image](https://user-images.githubusercontent.com/26014420/213017093-7f8cec85-c577-4e4f-93f7-91b52299e437.png)

Here's a result where I aligned to baseline but adjusted spacing of the bottom multiplicatively.  Bottom looks ok, but top is not aligned consistently with bottom.   This is with top aligned to baseline and baseline set tsize below the radius.  You do realize that tsize has no connection to the actual height of the text, right?  So aligning the top text to the edge may *require* a valign option.  But if you do that, the spacing is terrible:

![image](https://user-images.githubusercontent.com/26014420/213017722-96698a49-440d-41da-807d-110e4d1f21cd.png)

Unless you make a spacing multiplier adjustment to the top as well.  

![image](https://user-images.githubusercontent.com/26014420/213017883-c91d79f7-c4db-4be8-94c4-7852cadfa237.png)

So here I have a spacing factor of 1.35 on the top but 1.2 on the bottom.  It seemed like maybe the spacing could be computed from the curve curvature...but not if it varies like that.

**Comment by adrianVmariano** [comment]
I investigated the issue with orientation of the letters, and there was a subtle error in how the letter normals and spacing were computed.  So here's the result of an updated version (with some debugging lines).  

![image](https://user-images.githubusercontent.com/26014420/213059372-4aa18b03-23ae-46e6-a389-ba44413fdd14.png)

That's top aligned and then here's baseline:

![image](https://user-images.githubusercontent.com/26014420/213059472-eecd4274-7902-4b15-a854-248ff940e8a4.png)

And simply scaling the spacing creates a problem that the normals aren't centered on the glyphs any more:

![image](https://user-images.githubusercontent.com/26014420/213059703-2f213510-edba-4237-a1db-2d8f9a346213.png)

**Comment by alidaf** [comment]
Hi, Sorry for the late response, I was away yesterday but thank you for reopening. It looks like I've opened a can of worms but that you've made some progress anyway. Could you send a link or links to where the relevant code is so I could have a look. I doubt I'll be particularly useful at the moment but I can at least try to match your efforts.

**Comment by adrianVmariano** [comment]
I pushed my current debugging code state to my repo here: https://github.com/adrianVmariano/BOSL2 
I think you just need shapes3d.scad for this.  Note since the above plot I figured out how to make kerning work properly, but not spacing.  Not even sure that spacing makes sense.  It definitely does NOT make sense for regular linear text.  It may be the case that setting kern to a multiple of the font size would be a reasonable option.  

Here's the example with kern=3:

![image](https://user-images.githubusercontent.com/26014420/213161717-b9fcc292-4bd4-4465-a588-e50a0ee6ddea.png)

If you want the arrow uncomment the stroke() module in path_text().

**Comment by adrianVmariano** [comment]
I consulted with Jordan, who wrote the textmetrics code, and he didn't have any ideas about how to improve the text layout other than manual adjustment.  I think the multiplicative spacing parameter is a bad idea, so that will go away.  The only other idea I have is trying to add more alignment points.  Like if you could align to 5 units above baseline, say, would that be useful?  Then you'd know where the alignment level was, so you could, for example, align 5 units above baseline and draw on a circle 6 units inside your disk.  Maybe you get better spacing then?  And it's a known reference point.  Then you have to do trial an error on the alignment height instead of the kern value, though.  I don't know if that improves anything.

**Comment by adrianVmariano** [comment]
Another idea is to compute the curvature of the path and use that information somehow to help with the spacing.

**Comment by alidaf** [comment]
Hi Adrian,
Thanks for taking this up. I hope you aren't burning the midnight oil.
This all came about because I'm creating a template to produce customisable coasters with images and text. I gave up on this because the text around the edges looked awful, especially when I had text like in the example I showed where text at the bottom and top were both readable without rotating the coaster. 
Your realignment looks so much better than it did before so I think that will help enormously. The last thing for me is for the spacing to be consistent between the top and bottom text and that may be as simple as using the text size to adjust the spacing or kerning by considering an "effective" path that is offset from the original by the text size. In my example I adjusted the actual path but that affects the spacing making the top text look further apart than the bottom text. Maybe just a spacing factor to at least adjust it by eye, which looks exactly like what you demonstrated in your last example by adjusting the kern factor. So you may have already cracked that nugget!
It never ceases to amaze me how BOSL/BOSL2 is written as SCAD code considering the complexity of what some of the functions/modules do!

**Comment by adrianVmariano** [comment]
I augmented vertical alignment to allow things like "baseline+3" or "top-4".  I think that's about as powerful and flexible as I can make it.  Keep in mind that the only way to make a path offset from the original by the text size is to use text metrics to find out the text size.  The actual text size has nothing to do with the `size=` number you specify.  

I think this modification is probably complete at this point.  I pushed the latest to my fork.
